client/json-rpc/rpc_service.c

```diff
@@ -47,10 +49,166 @@
 #define RPC_METHOD_NOT_FOUND -32601
 #define RPC_INVALID_PARAMS -32603
 #define RPC_INTERNAL_ERROR -32693
+#define RPC_WHITE_ADDR_LEN          64
+#define RPC_WHITE_MAX               8
+
 
 const uint32_t RPC_SERVER_PORT = 7677; //default http json-rpc port 7677
 
-static int send_response(struct xdag_rpc_connection * conn, char *response) {
+struct item {
+    struct item *prev, *next;
+    struct in_addr addr;
+};
+
+struct item *g_rpc_white_host = NULL;
+
+static int rpc_host_addr_check_ipv4(const char *host){
+
+  int n[4];
+  char c[4];
+
+  if (!host){
+    return 0;
+  }
+
+  if (sscanf(host, "%d%c%d%c%d%c%d%c",
+             &n[0], &c[0], &n[1], &c[1],
+             &n[2], &c[2], &n[3], &c[3]) == 7)
+
+  {
+    int i;
+    for(i = 0; i < 3; ++i)
+      if (c[i] != '.')
+        return 0;
+    for(i = 0; i < 4; ++i)
+      if (n[i] > 255 || n[i] < 0)
+        return 0;
+    return 1;
+  } else
+    return 0;
+}
+
+static int rpc_white_host_check(struct sockaddr_in peeraddr){
+
+  struct item *node = NULL ;
+    
+  if (!g_rpc_white_host){
+        return 1;
+  }
+
+  LL_FOREACH(g_rpc_white_host,node)
+  {
+    if (node->addr.s_addr == peeraddr.sin_addr.s_addr){
+        return 1;
+    }
+  }
+
+  return 0;
+}
+
+
+static int rpc_white_host_add(const char *host){
+  struct item *new_white_host = NULL;
+  int white_num = 0;
+
+  if (!rpc_host_addr_check_ipv4(host)){
+    xdag_err("ip address is invalid");
+    return -1;
+  }
+
+  LL_COUNT(g_rpc_white_host,new_white_host,white_num);
+  if (white_num >= RPC_WHITE_MAX){
+    xdag_err("white list number is up to maximum");
+    return -2;
+  }
+
+  new_white_host = malloc(sizeof(struct item));
+  if (NULL == new_white_host){
+    xdag_err("memory is not enough.");
+    return -3;
+  }
+
+  new_white_host->addr.s_addr = inet_addr(host);
+  DL_APPEND(g_rpc_white_host, new_white_host);
+
+  return 0;
+}
+
+static int rpc_white_host_del(const char *host){
+
+  struct item *node = NULL ,*del_node = NULL;
+  struct in_addr addr = {0};
+
+  LL_FOREACH(g_rpc_white_host,node)
+  {
+    addr.s_addr = inet_addr(host);
+    if (addr.s_addr == node->addr.s_addr){
+        del_node = node;
+        LL_DELETE(g_rpc_white_host, del_node);
+        free(del_node);
+        return 0;
+    }
+  }
+
+  return -1;
+}
+
+static char  *rpc_white_host_query(){
+
+    static char result[RPC_WHITE_MAX * RPC_WHITE_ADDR_LEN] = {0};
+    struct item *node = NULL;
+    char new_host[RPC_WHITE_ADDR_LEN] = {0};
+
+    LL_FOREACH(g_rpc_white_host,node)
+    {
+        memset(new_host, 0, sizeof(new_host));
+        
+        sprintf(new_host, "[%s]",inet_ntoa(node->addr));
```

Comment:
@g2fly99  Multi line is better than single line.
[192.168.1.1]
[192.168.1.2]
[192.168.1.3]